mlflow/types/schema.py

```diff
@@ -1329,3 +1330,116 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.params)
+
+
+def convert_dataclass_to_schema(dataclass):
+    """
+    Create a Schema object from a dataclass. The dataclass should have type hints
+    for each field. Fields in the dataclass can be composed of basic types or other dataclasses.
+    Only the top-level fields become ColSpecs directly. Lower-level fields are converted into
+    nested Object types.
+    """
+    field_type_mapping = {
+        str: "string",
+        int: "integer",
+        bool: "boolean",
+    }
+
+    inputs = []
+
+    for field_name, field_type in dataclass.__annotations__.items():
+        # Determine the type and handle Optional and List correctly
+        is_optional = False
+        effective_type = field_type
+
+        if get_origin(field_type) == Union and type(None) in get_args(field_type):
+            # This is an Optional type; determine the effective type excluding None
+            is_optional = True
+            effective_type = next(t for t in get_args(field_type) if t is not type(None))
+
+        if get_origin(effective_type) == list:
+            # It's a list, check the type within the list
+            list_type = get_args(effective_type)[0]
```

Comment:
Is it possible list_type is Union again?

Reply:
Added a check to disallow Union types (besides Optional)